### PR TITLE
Backwards Compatibility System [1.1.1]

### PR DESCRIPTION
**Pull request by larryk85**
With current DUNE, it will fail on old versions of nodeos being run.

### CODE REVIEW

File: src/dune/docker.py [hunk 1]
```diff
@@ -97,6 +97,18 @@ def cp_from_host(self, host_file, container_file):
...
+        return self.execute_cmd(['echo', body, '>', file_name])
+
+    def append_file(self, file_name, body):
+        return self.execute_cmd(['echo', body, '>>', file_name])
+
+    def get_arch(self) :
+        stdout, stderr, exit_code = self.execute_cmd(['uname', '-m'])
+        if stdout in ('x86_64\n', 'amd64\n') :
+            return 'amd64'
+        return 'arm64'
```
Comment (ScottBailey) [inline src/dune/docker.py:110]: anything that's not intel or amd 64 is arm? I don't have a better suggestion so it is what it is.

Comment (larryk85) [reply src/dune/docker.py:110]: Yeah, I'm not super enthused about it either. But, I couldn't prove to myself that all ARM systems would return aarch64, I read that some return arm8, arm9, etc.  I believe it should also be aarch64 for 64bit ARM, but once again I couldn't find definitive information and I didn't want to go off of the one Mac Mini as a 'proof' that that works.

Plus, we explicitly own the packages for these and only support arm64 or amd64.

File: src/dune/dune.py [hunk 2]
```diff
@@ -101,6 +103,10 @@ def create_node(self, nod):
         print("Creating node [" + nod.name() + "]")
         self._docker.execute_cmd(['mkdir', '-p', nod.data_dir()])
 
+    def get_current_nodeos_version(self):
+        stdout, stderr, exit_code = self._docker.execute_cmd(['nodeos', '--version'])
+        return stdout[1:].split('.')
```
Comment (ScottBailey) [inline src/dune/dune.py:108]: seems like either a regex or simply taking the whole string might future proof this a bit better.

File: src/dune/dune.py [hunk 3]
```diff
@@ -123,6 +129,13 @@ def start_node(self, nod, snapshot=None):
 
         # copy config.ini to config-dir
         if not is_restart and nod.config() is None:
+            current_ver = self.get_current_nodeos_version()
+            config_ini = get_config_ini(self._docker.get_arch(), current_ver[0], current_ver[1], current_ver[2])
```
Comment (ScottBailey) [inline src/dune/dune.py:133]: consider:
```suggestion
            config_ini = get_config_ini(self._docker.get_arch(), *current_ver)
```

File: src/dune/dune.py (hunk 2 above)
Comment (larryk85) [reply src/dune/dune.py:108]: Good point.

Comment (ScottBailey) [reply src/dune/dune.py:108]: for example, what do we do with the trailing release info? e.g. `v4.1.0-dev`. Should patch really be `0-dev`?

Comment (ScottBailey) [inline src/dune/dune.py:108]:
```suggestion
        return stdout[1:].split('.-')
```

File: src/dune/dune.py [hunk 4]
```diff
@@ -123,6 +129,13 @@ def start_node(self, nod, snapshot=None):
...
+            config_file = tempfile.NamedTemporaryFile("w+", encoding="UTF-8", delete=False)
+            config_file.write(config_ini)
+            config_file.close()
+            self._docker.cp_from_host(config_file.name, "/app/config.ini")
```
Comment (ScottBailey) [inline src/dune/dune.py:137]: Would error checking be appropriate here?

File: src/dune/dune.py (hunk 2 above)
Comment (larryk85) [reply src/dune/dune.py:108]: Thank you, completely forgot about the occasional suffix.

Comment (ScottBailey) [reply src/dune/dune.py]: For future, we probably should split on '.-+'. Forgot about build suffixes until just now.

**Comment by ScottBailey** [comment]
test/bootstrap_test.py fails.

```
warn  2023-05-19T20:51:25.689 nodeos    producer_plugin.cpp:1216      plugin_initialize    ] 3110006 plugin_config_exception: Incorrect plugin configuration
--read-only-read-window-time-us (60000) must be greater than --max-transaction-time 1000 ms plus a margin of 10000 us
    {"read":60000,"trx_time":1000,"min":10000}
    nodeos  producer_plugin.cpp:1174 plugin_initialize

error 2023-05-19T20:51:25.692 nodeos    main.cpp:161                  main                 ] 3110006 plugin_config_exception: Incorrect plugin configuration
--read-only-read-window-time-us (60000) must be greater than --max-transaction-time 1000 ms plus a margin of 10000 us
    {"read":60000,"trx_time":1000,"min":10000}
    nodeos  producer_plugin.cpp:1174 plugin_initialize
rethrow
    {}
    nodeos  producer_plugin.cpp:1216 plugin_initialize


Creating account [myaccount] with key pair [Private: 5KEZDdpkAbiHfv6gb5iCWgx9UaS12dUPJ8W6Z2CNefeSoWuLrmo, Public: EOS53t6rPR3NDjY8T86SbSVtSjmktdiimHZRayZXoHtZXdhPLvdxA]
```

**Review by ScottBailey — CHANGES_REQUESTED** [review]
booststrao_test.py fails

**Comment by mikelik** [comment]
I have added some bug fixes and pylint fixes in https://github.com/AntelopeIO/DUNE/pull/185

**Review by jolly-fellow — APPROVED** [review]

**Review by mikelik — CHANGES_REQUESTED** [review]
Revert port change

**Review by mikelik — APPROVED** [review]

**Review by ScottBailey — APPROVED** [review]
lgtm

**Pull request merged**